### PR TITLE
contrib/gocql/gocql: support Scanner and Batch

### PR DESCRIPTION
**Pull request by jack-at-circle**
I lost my first commit somehow, but it got included in the second one... 

Adds:
 - Support for calling Iter.Scanner() in such a way that the Iter's trace actually closes
 - Support for Batch queries including calling WithContext and adding Query's to the batch after it has been wrapped
 - `batch.ExecuteBatch(session) error` function which inverts the `session.ExecuteBatch(batch) error` signature to execute the batch with tracing
 - Test coverage

Additionally: 
WrapQuery was incorrectly using context.Background() instead of the passed query's Context method that could already have a parent trace attached to it.

### CODE REVIEW

File: contrib/gocql/gocql/gocql.go [hunk 1]
```diff
@@ -163,3 +176,92 @@ func (tIter *Iter) Close() error {
...
+	}
+}
+
+// Err calls the wrapped Scanner.Err, releasing both Scanner and Iter resources.
+func (s *Scanner) Err() error {
+	if err := s.Iter.Close(); err != nil {
+		s.Scanner.Err()
+		return err
+	}
+
```
Comment (gbbr) [inline contrib/gocql/gocql/gocql.go:196]: nit: can you please remove this empty line?

File: contrib/gocql/gocql/gocql.go [hunk 2]
```diff
@@ -163,3 +176,92 @@ func (tIter *Iter) Close() error {
...
+	return &Scanner{
+		Scanner: tIter.Iter.Scanner(),
+		Iter:    tIter,
+	}
+}
+
+// Err calls the wrapped Scanner.Err, releasing both Scanner and Iter resources.
+func (s *Scanner) Err() error {
+	if err := s.Iter.Close(); err != nil {
+		s.Scanner.Err()
```
Comment (gbbr) [inline contrib/gocql/gocql/gocql.go:193]: Why is this call here? It's a bit odd and I wonder if it warrants a comment explaining the reason.

File: contrib/gocql/gocql/gocql_test.go [hunk 3]
```diff
@@ -103,9 +103,9 @@ func TestChildWrapperSpan(t *testing.T) {
 	cluster := newCassandraCluster()
 	session, err := cluster.CreateSession()
 	assert.Nil(err)
-	q := session.Query("SELECT * from trace.person")
+	q := session.Query("SELECT * from trace.person").WithContext(ctx)
 	tq := WrapQuery(q, WithServiceName("TestServiceName"))
-	iter := tq.WithContext(ctx).Iter()
```
Comment (gbbr) [inline contrib/gocql/gocql/gocql_test.go:108]: Why this change?

**Review by gbbr — COMMENTED** [review]
Very nicely done! Just a few questions, but otherwise good to go! :shipit:

File: contrib/gocql/gocql/gocql_test.go (hunk 3 above)
Comment (jack-at-circle) [reply contrib/gocql/gocql/gocql_test.go:108]: I'll add a comment, but this test change proves the need for changing the default context of a Query from context.Background to query.Context because WithContext could have already been called on the query before WrapQuery is called. i.e. the parent trace could get lost depending on the order of calls before this change and the test proves it.

File: contrib/gocql/gocql/gocql.go (hunk 2 above)
Comment (jack-at-circle) [reply contrib/gocql/gocql/gocql.go]: I realized this implementation was an artifact of how I initially wrapped the package. The most recent commit mirrors the Iter.Close function directly, which is cleaner.

**Review by gbbr — APPROVED** [review]
Very nice job! Thank you! Please rebase on the `v1` branch and we should be good to merge!

**Comment by jack-at-circle** [comment]
> Very nice job! Thank you! Please rebase on the `v1` branch and we should be good to merge!

I tried rebasing... looks like it didn't work. This is the first time I've merged from a fork, sorry!

**Comment by jack-at-circle** [comment]
> Very nice job! Thank you! Please rebase on the `v1` branch and we should be good to merge!

Sorry! I somehow did that very wrong and both deleted the branch and the commits... first time trying to rebase across forks... I'll open another pr tomorrow
